### PR TITLE
fix(dms): update the API for resizing RabbitMQ and Kafka instances

### LINKED ISSUE #2675: [DMS/Kafka] The `flavor_id` argument does not support changes in `huaweicloud_dms_kafka_instance`
### Terraform Version

```
1.44.1
```

After modifying the `flavor_id` argument, the kafka instance specification does not change.

### PR DESCRIPTION
**Pull request by chengxiangdong**
**What this PR does / why we need it**:

- Use RabbitMQ's API to resize RabbitMQ instances.
- Use Kafka's API to resize Kafka instances.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #2675

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
=== RUN   TestAccDmsRabbitmqInstances_basic
=== PAUSE TestAccDmsRabbitmqInstances_basic
=== CONT  TestAccDmsRabbitmqInstances_basic
=== RUN   TestAccDmsRabbitmqInstances_withEpsId
=== PAUSE TestAccDmsRabbitmqInstances_withEpsId
=== CONT  TestAccDmsRabbitmqInstances_withEpsId
=== RUN   TestAccDmsRabbitmqInstances_compatible
=== PAUSE TestAccDmsRabbitmqInstances_compatible
=== CONT  TestAccDmsRabbitmqInstances_compatible
--- PASS: TestAccDmsRabbitmqInstances_compatible (1114.07s)
=== RUN   TestAccDmsRabbitmqInstances_single
=== PAUSE TestAccDmsRabbitmqInstances_single
=== CONT  TestAccDmsRabbitmqInstances_single
--- PASS: TestAccDmsRabbitmqInstances_single (505.12s)
--- PASS: TestAccDmsRabbitmqInstances_withEpsId (2247.60s)
--- PASS: TestAccDmsRabbitmqInstances_basic (2571.71s)
PASS
```

```
=== RUN   TestAccKafkaInstance_basic
=== PAUSE TestAccKafkaInstance_basic
=== CONT  TestAccKafkaInstance_basic
=== CONT  TestAccKafkaInstance_basic
--- PASS: TestAccKafkaInstance_basic (885.70s)
=== RUN   TestAccKafkaInstance_withEpsId
=== PAUSE TestAccKafkaInstance_withEpsId
=== CONT  TestAccKafkaInstance_withEpsId
--- PASS: TestAccKafkaInstance_withEpsId (715.33s)
=== RUN   TestAccKafkaInstance_compatible
=== PAUSE TestAccKafkaInstance_compatible
=== CONT  TestAccKafkaInstance_compatible
--- PASS: TestAccKafkaInstance_compatible (811.28s)
=== RUN   TestAccKafkaInstance_newFormat
=== PAUSE TestAccKafkaInstance_newFormat
=== CONT  TestAccKafkaInstance_newFormat
--- PASS: TestAccKafkaInstance_newFormat (2227.86s)
PASS
```

### CODE REVIEW

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go [hunk 1]
```diff
@@ -833,68 +825,141 @@ func resourceDmsKafkaInstanceUpdate(ctx context.Context, d *schema.ResourceData,
...
-		flavorId = d.Get("flavor_id").(string)
+	if d.HasChanges("broker_num") {
+		operType := "horizontal"
+		brokerNum := d.Get("broker_num").(int)
+
+		resizeOpts := instances.ResizeInstanceOpts{
+			OperType:     &operType,
+			NewBrokerNum: &brokerNum,
+		}
+		log.Printf("[DEBUG] Resize Kafka instance flavor ID options: %#v", resizeOpts)
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go]: log.Printf("[DEBUG] Resize Kafka instance flavor ID options: %#v", resizeOpts)
log.Printf("[DEBUG] Resize Kafka instance broker number options: %#v", resizeOpts)

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go [hunk 2]
```diff
@@ -833,68 +825,141 @@ func resourceDmsKafkaInstanceUpdate(ctx context.Context, d *schema.ResourceData,
...
+		product, err := getKafkaProductDetails(cfg, d)
+		if err != nil {
+			return fmt.Errorf("failed to resize Kafka instance, query product details error: %s", err)
+		}
+		storageSpace := d.Get("storage_space").(int)
+		resizeOpts := instances.ResizeInstanceOpts{
+			NewSpecCode:     &product.SpecCode,
+			NewStorageSpace: &storageSpace,
+		}
+		log.Printf("[DEBUG] Resize Kafka instance storage space options: %#v", resizeOpts)
```
Comment (houpeng80) [inline huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go:845]: it is better to describe the product info in the log info

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go (hunk 1 above)
Comment (chengxiangdong) [reply huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go [hunk 3]
```diff
@@ -833,68 +825,141 @@ func resourceDmsKafkaInstanceUpdate(ctx context.Context, d *schema.ResourceData,
...
+			return v, "PENDING", nil
+		}
+
+		if operType != nil && *operType == "storage" && v.TotalStorageSpace != storageSpace {
+			return v, "PENDING", nil
+		}
+
+		if operType != nil && *operType == "horizontal" && v.BrokerNum != brokerNum {
+			return v, "PENDING", nil
+		}
```
Comment (Lance52259) [inline huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go]:
```suggestion
		if ((operType == nil || *operType == "vertical") && v.ProductID != flavorID) || 
			(operType != nil && *operType == "storage" && v.TotalStorageSpace != storageSpace) ||
				(operType != nil && *operType == "horizontal" && v.BrokerNum != brokerNum) {
			return v, "PENDING", nil
		}
```

Comment (chengxiangdong) [reply huaweicloud/services/dms/resource_huaweicloud_dms_kafka_instance.go]: done

**Pull request merged**